packages/core/src/storages/request_queue_v2.ts

```diff
@@ -0,0 +1,988 @@
+    Dictionary,
+    ProcessedRequest,
+    QueueOperationInfo,
+    RequestQueueClient,
+    RequestQueueInfo,
+    StorageClient,
+} from '@crawlee/types';
+import type { DownloadListOfUrlsOptions } from '@crawlee/utils';
+import { chunk, downloadListOfUrls, sleep } from '@crawlee/utils';
+import ow from 'ow';
+
+import type { AddRequestsBatchedOptions, AddRequestsBatchedResult } from './request_queue';
+import type { StorageManagerOptions } from './storage_manager';
+import { StorageManager } from './storage_manager';
+import {
+    API_PROCESSED_REQUESTS_DELAY_MILLIS,
+    MAX_QUERIES_FOR_CONSISTENCY,
+    QUERY_HEAD_BUFFER,
+    QUERY_HEAD_MIN_LENGTH,
+    STORAGE_CONSISTENCY_DELAY_MILLIS,
+    getRequestId,
+    purgeDefaultStorages,
+} from './utils';
+import { Configuration } from '../configuration';
+import { log } from '../log';
+import type { ProxyConfiguration } from '../proxy_configuration';
+import { Request } from '../request';
+import type { InternalSource, RequestOptions, Source } from '../request';
+
+// Double the limit of RequestQueue v1 (1_000_000) as we also store keyed by request.id, not just from uniqueKey
+const MAX_CACHED_REQUESTS = 2_000_000;
+
+/**
+ * When prolonging a lock, we do it for a minute from Date.now()
+ */
+const PROLONG_LOCK_BY_SECS = 60;
+
+/**
+ * This number must be large enough so that processing of all these requests cannot be done in
+ * a time lower than expected maximum latency of DynamoDB, but low enough not to waste too much memory.
+ * @internal
+ */
+const RECENTLY_HANDLED_CACHE_SIZE = 1000;
+
+interface RequestLruItem {
+    uniqueKey: string;
+    isHandled: boolean;
+    id: string;
+    hydrated: Request | null;
+    lockExpiresAt: number | null;
+}
+
+export interface RequestQueueV2OperationOptions {
+    /**
+     * If set to `true`:
+     *   - while adding the request to the queue: the request will be added to the foremost position in the queue.
+     *   - while reclaiming the request: the request will be placed to the beginning of the queue, so that it's returned
+     *   in the next call to {@apilink RequestQueueV2.fetchNextRequest}.
+     * By default, it's put to the end of the queue.
+     * @default false
+     */
+    forefront?: boolean;
+}
+
+/**
+ * @internal
+ */
+export interface RequestQueueV2OperationInfo extends QueueOperationInfo {
+    uniqueKey: string;
+}
+
+class RequestQueue {
+    log = log.child({ prefix: 'RequestQueueV2' });
+    id: string;
+    name?: string;
+    timeoutSecs = 30;
+    clientKey = cryptoRandomObjectId();
+    client: RequestQueueClient;
+    private proxyConfiguration?: ProxyConfiguration;
+
+    private queueHeadIds = new ListDictionary<string>();
+    private requestCache = new LruCache<RequestLruItem>({ maxLength: MAX_CACHED_REQUESTS });
+    private requestIdsInProgress = new Set<string>();
+    private recentlyHandledRequestsCache = new LruCache<boolean>({ maxLength: RECENTLY_HANDLED_CACHE_SIZE });
+
+    lastActivity = new Date();
```

Comment:
this was added to the v1 to overcome the 0 concurrency issues, v2 shouldn't really suffer from those, as a stale request should be automatically released and retried.

but lets keet it here for now I guess to be on the safe side. maybe just add a comment that this shouldn't be needed.